staging/src/k8s.io/apiserver/pkg/server/options/etcd.go

```diff
@@ -223,12 +228,58 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
 	}
 
 	if len(s.EncryptionProviderConfigFilepath) != 0 {
-		transformerOverrides, kmsPluginHealthzChecks, err := encryptionconfig.LoadEncryptionConfig(s.EncryptionProviderConfigFilepath, s.EncryptionProviderConfigAutomaticReload, stopCh)
+		ctx, closeTransformers := wait.ContextForChannel(stopCh)
+
+		encryptionConfiguration, err := encryptionconfig.LoadEncryptionConfig(s.EncryptionProviderConfigFilepath, s.EncryptionProviderConfigAutomaticReload, ctx.Done())
 		if err != nil {
+			// in case of error, we want to close partially initialized (if any) transformers
+			closeTransformers()
 			return err
 		}
-		s.transformerOverrides = transformerOverrides
-		s.kmsPluginHealthzChecks = kmsPluginHealthzChecks
+
+		s.kmsPluginHealthzChecks = encryptionConfiguration.HealthChecks
+
+		var kmsPluginHealthzCheck healthz.HealthChecker
+		if len(encryptionConfiguration.HealthChecks) > 0 {
+			kmsPluginHealthzCheck = encryptionConfiguration.HealthChecks[0]
+		}
```

Comment:
it's confusing to propagate this into NewDynamicTransformers when it's not going to be used in the non-reload case... I'm trying to think how to make it clearer that there's ~two modes NewDynamicTransformers operates in

* non-reloading: it is not a health checker, Set will never be called, it's only handling delegating transformer lookups for resources. we could construct with `encryptionconfig.NewDynamicTransformers(encryptionConfiguration.Transformers)`
* reloading: it *is* a health checker, Set will be called. we could construct with `encryptionconfig.NewReloadingDynamicTransformers(encryptionConfiguration.Transformers, kmsPluginHealthzCheck, closeTransformers, encryptionConfiguration.KMSTimeout)`